File: android/app/src/main/java/com/lilru/liftr/territory/TerritoryOwnerColors.kt
package com.lilru.liftr.territory

import androidx.compose.ui.graphics.Color
import java.util.UUID

object TerritoryOwnerColors {
  private val palette: List<Color> = (0 until 48).map { index ->
    val hue = index / 48f * 360f
    Color.hsv(hue, 0.72f, 0.88f)
  }

  fun color(ownerId: String?): Color {
    return palette[paletteIndex(ownerId)]
  }

  fun fill(ownerId: String?, isMine: Boolean): Color {
    return color(ownerId).copy(alpha = if (isMine) 0.52f else 0.28f)
  }

  fun stroke(ownerId: String?, isMine: Boolean): Color {
    val base = color(ownerId)
    return if (isMine) base else base.copy(alpha = 0.82f)
  }

  fun strokeWidth(isMine: Boolean): Float = if (isMine) 2.5f else 1f

  private fun paletteIndex(ownerId: String?): Int {
    if (ownerId.isNullOrBlank()) return 0
    return runCatching {
      val uuid = UUID.fromString(ownerId)
      val bytes = ByteArray(16)
      val msb = uuid.mostSignificantBits
      val lsb = uuid.leastSignificantBits
      for (i in 0 until 8) {
        bytes[i] = ((msb shr (56 - i * 8)) and 0xff).toByte()
        bytes[i + 8] = ((lsb shr (56 - i * 8)) and 0xff).toByte()
      }
      var hash = 5381
      for (byte in bytes) {
        hash = ((hash shl 5) + hash) + (byte.toInt() and 0xFF)
      }
      var secondary = 0
      for (byte in bytes.reversed()) {
        secondary = ((secondary shl 5) + secondary) + (byte.toInt() and 0xFF)
      }
      kotlin.math.abs(hash + secondary * 17) % palette.size
    }.getOrDefault(ownerId.hashCode().let { if (it < 0) -it else it } % palette.size)
  }
}

File: Liftr/TerritoryOwnerColors.swift
import SwiftUI
import UIKit

enum TerritoryOwnerColors {
    private static let palette: [Color] = (0..<48).map { index in
        let hue = Double(index) / 48.0
        return Color(hue: hue, saturation: 0.72, brightness: 0.88)
    }

    private static func paletteIndex(for ownerId: UUID) -> Int {
        var hash = 5381
        let bytes = withUnsafeBytes(of: ownerId.uuid) { Array($0) }
        for byte in bytes {
            hash = ((hash << 5) &+ hash) &+ Int(byte)
        }
        var secondary = 0
        for byte in bytes.reversed() {
            secondary = ((secondary << 5) &+ secondary) &+ Int(byte)
        }
        return abs(hash &+ secondary &* 17) % palette.count
    }

    static func color(for ownerId: UUID) -> Color {
        palette[paletteIndex(for: ownerId)]
    }

    static func fill(for ownerId: UUID, isMine: Bool) -> Color {
        color(for: ownerId).opacity(isMine ? 0.50 : 0.44)
    }

    static func uiColor(for ownerId: UUID) -> UIColor {
        UIColor(
            hue: CGFloat(Double(paletteIndex(for: ownerId)) / 48.0),
            saturation: 0.72,
            brightness: 0.88,
            alpha: 1.0
        )
    }

    static func uiFill(for ownerId: UUID, isMine: Bool) -> UIColor {
        uiColor(for: ownerId).withAlphaComponent(isMine ? 0.50 : 0.44)
    }

    static func stroke(for ownerId: UUID, isMine: Bool, denseOverlay: Bool = false) -> Color {
        if denseOverlay {
            return .clear
        }
        let base = color(for: ownerId)
        return isMine ? base : base.opacity(0.82)
    }

    static func strokeWidth(isMine: Bool, denseOverlay: Bool = false) -> CGFloat {
        if denseOverlay {
            return 0
        }
        return isMine ? 3.0 : 1.0
    }
}
